File: src/cli/bundle/html-plugin.ts
import fs from 'node:fs/promises'
import path from 'node:path'
import type esbuild from 'esbuild'
import type {Config} from '../types/config.ts'
import {parseHTML} from '../utils/html-parser.ts'

type Manifest = {
  icons?: {src?: string; size?: string; type?: string}[]
  start_url?: string
  version?: string
}

// to-do: a generic utility that recursively inlines all resources in an HTML
//        file.
export function HTMLPlugin(config: Readonly<Config>): esbuild.Plugin {
  return {
    name: 'HTMLPlugin',
    setup(build) {
      build.onEnd(async _result => {
        const doc = await parseHTML(config.entry)

        const scripts = doc.querySelectorAll<HTMLScriptElement>(
          "script[type='module'][src$='.ts']"
        )
        for (const script of scripts) {
          // to-do: test.
          const filename = path.basename(
            script.getAttribute('src')!.replace(/\.ts$/, '.js')
          )
          if (config.oneFile) {
            script.removeAttribute('src')
            script.textContent = await fs.readFile(
              path.join(config.out.dir, filename),
              {encoding: 'utf8'}
            )
          } else script.src = filename
        }

        if (config.oneFile) {
          const imgs = doc.querySelectorAll<HTMLImageElement>('img[src]')
          for (const img of imgs) {
            const filename = img.getAttribute('src')!
            const b64 = await fs.readFile(path.join(config.out.dir, filename), {
              encoding: 'base64'
            })
            const ext = filename.match(/\.([^.]+)$/)?.[1]
            if (!ext) throw Error('no asset extension')
            img.src = `data:image/${ext};base64,${b64}`
          }
        }

        const manifestEl = doc.querySelector<HTMLLinkElement>(
          'link[href][rel="manifest"]'
        )
        if (manifestEl) {
          const manifestFilename = path.resolve(
            path.dirname(config.entry),
            manifestEl.getAttribute('href')!
          )
          const manifest: Manifest = JSON.parse(
            await fs.readFile(manifestFilename, 'utf8')
          )
          if (config.bundle.version) manifest.version = config.bundle.version

          for (const icon of manifest.icons ?? []) {
            if (!icon.src || !icon.type) continue
            const iconFilename = `${path.dirname(manifestFilename)}/${icon.src}`
            const file = await fs.readFile(iconFilename)
            if (config.oneFile)
              icon.src = `data:${icon.type};base64,${file.toString('base64')}`
            else icon.src = path.relative(config.out.dir, iconFilename)
          }
          if (config.watch)
            manifest.start_url = `http://localhost:${config.port}`

          if (config.oneFile)
            manifestEl.href = `data:application/json,${encodeURIComponent(
              JSON.stringify(manifest)
            )}`
          else
            await fs.writeFile(
              path.join(config.out.dir, path.basename(manifestFilename)),
              JSON.stringify(manifest)
            )
        }

        if (config.oneFile) {
          const icons = doc.querySelectorAll<HTMLLinkElement>(
            'link[href][rel="icon"][type]'
          )
          for (const icon of icons) {
            const filename = path.resolve(
              config.out.dir,
              icon.getAttribute('href')!
            )
            const file = await fs.readFile(filename)
            icon.href = `data:${icon.type};base64,${file.toString('base64')}`
          }
        }

        // to-do: test.
        const html = `<!doctype html>\n${doc.documentElement.outerHTML}`
        const outHTMLFilename = path.join(config.out.dir, config.out.filename)
        await fs.writeFile(outHTMLFilename, html)
      })
    }
  }
}

File: src/cmd/pack/plugins/html_plugin.go
package plugins

import (
	"encoding/base64"
	"encoding/json"
	"mime"
	"os"
	"path/filepath"
	"strings"

	"github.com/evanw/esbuild/pkg/api"
	"github.com/oidoid/void/src/cmd/pack/cliconfig"
	"github.com/oidoid/void/src/cmd/pack/htmlparser"
	"golang.org/x/net/html"
)

func HTMLPlugin(config *cliconfig.CLIConfig) api.Plugin {
	return api.Plugin{
		Name: "HTMLPlugin",
		Setup: func(build api.PluginBuild) {
			build.OnEnd(func(result *api.BuildResult) (api.OnEndResult, error) {
				if len(result.Errors) > 0 {
					return api.OnEndResult{}, nil
				}
				for _, entry := range config.Entries {
					if filepath.Ext(entry) != ".html" {
						continue
					}
					if result := transformHTML(config, entry, filepath.Dir(entry)); len(result.Errors) > 0 {
						return result, nil
					}
				}
				return api.OnEndResult{}, nil
			})
		},
	}
}

func transformHTML(config *cliconfig.CLIConfig, entry, entryDir string) api.OnEndResult {
	doc, err := htmlparser.ParseDoc(entry)
	if err != nil {
		return errorEndResult(err)
	}
	if result := transformFavicons(doc, config, entryDir); len(result.Errors) > 0 {
		return result
	}
	if result := transformImages(doc, config, entryDir); len(result.Errors) > 0 {
		return result
	}
	if result := transformManifests(doc, config, entryDir); len(result.Errors) > 0 {
		return result
	}
	if result := transformScripts(doc, config); len(result.Errors) > 0 {
		return result
	}
	if result := transformStylesheets(doc, config); len(result.Errors) > 0 {
		return result
	}
	filename := filepath.Join(config.OutDir, filepath.Base(entry))
	file, err := os.Create(filename)
	if err != nil {
		return errorEndResult(err)
	}
	defer file.Close()
	if err := html.Render(file, doc); err != nil {
		return errorEndResult(err)
	}
	return api.OnEndResult{}
}

func transformFavicons(doc *html.Node, config *cliconfig.CLIConfig, entryDir string) api.OnEndResult {
	for _, node := range htmlparser.QueryTag(doc, "link") {
		rel := htmlparser.NodeAttr(node, "rel")
		if rel != "icon" {
			continue
		}
		href := htmlparser.NodeAttr(node, "href")
		if href == "" {
			continue
		}
		src, err := transformImageSrc(config, entryDir, href)
		if err != nil {
			return errorEndResult(err)
		}
		htmlparser.SetNodeAttr(node, "href", src)
	}
	return api.OnEndResult{}
}

func transformImages(doc *html.Node, config *cliconfig.CLIConfig, entryDir string) api.OnEndResult {
	for _, node := range htmlparser.QueryTag(doc, "img") {
		src := htmlparser.NodeAttr(node, "src")
		if src == "" {
			continue
		}
		newSrc, err := transformImageSrc(config, entryDir, src)
		if err != nil {
			return errorEndResult(err)
		}
		htmlparser.SetNodeAttr(node, "src", newSrc)
	}
	return api.OnEndResult{}
}

func transformImageSrc(config *cliconfig.CLIConfig, entryDir, src string) (string, error) {
	filename := filepath.Base(src)
	bin, err := os.ReadFile(filepath.Join(entryDir, src))
	if err != nil {
		return "", err
	}
	if config.OneFile {
		mimeType := mime.TypeByExtension(strings.ToLower(filepath.Ext(filename)))
		if mimeType == "" {
			mimeType = "application/octet-stream"
		}
		return "data:" + mimeType + ";base64," + base64.StdEncoding.EncodeToString(bin), nil
	}
	if err := os.WriteFile(filepath.Join(config.OutDir, filename), bin, 0o644); err != nil {
		return "", err
	}
	return filename, nil
}

func transformManifests(doc *html.Node, config *cliconfig.CLIConfig, entryDir string) api.OnEndResult {
	for _, node := range htmlparser.QueryTag(doc, "link") {
		if htmlparser.NodeAttr(node, "rel") != "manifest" {
			continue
		}
		href := htmlparser.NodeAttr(node, "href")
		if href == "" {
			continue
		}
		filename := filepath.Base(href)
		srcBin, err := os.ReadFile(filepath.Join(entryDir, href))
		if err != nil {
			return errorEndResult(err)
		}

		var manifest map[string]any
		if err := json.Unmarshal(srcBin, &manifest); err != nil {
			return errorEndResult(err)
		}
		if icons, ok := manifest["icons"].([]any); ok {
			for _, item := range icons {
				icon, ok := item.(map[string]any)
				if !ok {
					continue
				}
				iconSrc := icon["src"].(string)
				if iconSrc == "" {
					continue
				}
				newSrc, err := transformImageSrc(config, entryDir, iconSrc)
				if err != nil {
					return errorEndResult(err)
				}
				icon["src"] = newSrc
			}
		}

		// to-do:
		// if (config.bundle.version) manifest.version = config.bundle.version
		// if (config.watch)
		//   manifest.start_url = `http://localhost:${config.port}`

		manifestBin, err := json.Marshal(manifest)
		if err != nil {
			return errorEndResult(err)
		}
		if config.OneFile {
			dataURI := "data:application/manifest+json;base64," + base64.StdEncoding.EncodeToString(manifestBin)
			htmlparser.SetNodeAttr(node, "href", dataURI)
			continue
		}
		if err := os.WriteFile(filepath.Join(config.OutDir, filename), manifestBin, 0o644); err != nil {
			return errorEndResult(err)
		}
		htmlparser.SetNodeAttr(node, "href", filename)
	}
	return api.OnEndResult{}
}

func transformScripts(doc *html.Node, config *cliconfig.CLIConfig) api.OnEndResult {
	for _, node := range htmlparser.QueryTag(doc, "script") {
		src := htmlparser.NodeAttr(node, "src")
		if src == "" || htmlparser.NodeAttr(node, "type") != "module" {
			continue
		}
		filename := filepath.Base(src)
		if strings.HasSuffix(filename, ".ts") {
			filename = filename[:len(filename)-3] + ".js"
		}

		if config.OneFile {
			bin, err := os.ReadFile(filepath.Join(config.OutDir, filename))
			if err != nil {
				return errorEndResult(err)
			}
			htmlparser.RemoveNodeAttr(node, "src")
			htmlparser.SetTextContent(node, string(bin))
			continue
		}

		htmlparser.SetNodeAttr(node, "src", filename)
	}
	return api.OnEndResult{}
}

func transformStylesheets(doc *html.Node, config *cliconfig.CLIConfig) api.OnEndResult {
	for _, node := range htmlparser.QueryTag(doc, "link") {
		if htmlparser.NodeAttr(node, "rel") != "stylesheet" {
			continue
		}
		href := htmlparser.NodeAttr(node, "href")
		if href == "" {
			continue
		}
		filename := filepath.Base(href)

		bin, err := os.ReadFile(filepath.Join(config.OutDir, filename))
		if err != nil {
			return errorEndResult(err)
		}
		if config.OneFile {
			styleNode := &html.Node{Type: html.ElementNode, Data: "style"}
			htmlparser.SetTextContent(styleNode, string(bin))
			htmlparser.ReplaceNode(node, styleNode)
			continue
		}
		htmlparser.SetNodeAttr(node, "href", filename)
	}
	return api.OnEndResult{}
}
